### PR TITLE
Remove warnings from -Wchar-subscripts for known positive constants

### PR DESCRIPTION
**Pull request by wheatman**
This is to address https://github.com/llvm/llvm-project/issues/18763

it removes warnings from using a signed char as an array bound if the char is a known positives constant.

This goes one step farther than gcc does.

For example given the following code
```c++
char upper[300];

int main() {
  upper['a'] = 'A';
  char b = 'a';
  upper[b] = 'A';
  const char c = 'a';
  upper[c] = 'A';
  constexpr char d = 'a';
  upper[d] = 'A';
  char e = -1;
  upper[e] = 'A';
  const char f = -1;
  upper[f] = 'A';
  constexpr char g = -1;
  upper[g] = 'A';
  return 1;
}
```

clang currently gives warnings for all cases, while gcc gives warnings for all cases except for 'a' (https://godbolt.org/z/5ahjETTv3)

with the change there is no longer any warning for 'a', 'c', or 'd'.

```
../test.cpp:7:8: warning: array subscript is of type 'char' [-Wchar-subscripts]
    7 |   upper[b] = 'A';
      |        ^~
../test.cpp:13:8: warning: array subscript is of type 'char' [-Wchar-subscripts]
   13 |   upper[e] = 'A';
      |        ^~
../test.cpp:15:8: warning: array subscript is of type 'char' [-Wchar-subscripts]
   15 |   upper[f] = 'A';
      |        ^~
../test.cpp:17:8: warning: array subscript is of type 'char' [-Wchar-subscripts]
   17 |   upper[g] = 'A';
      |        ^~
../test.cpp:15:3: warning: array index -1 is before the beginning of the array [-Warray-bounds]
   15 |   upper[f] = 'A';
      |   ^     ~
../test.cpp:1:1: note: array 'upper' declared here
    1 | char upper[300];
      | ^
../test.cpp:17:3: warning: array index -1 is before the beginning of the array [-Warray-bounds]
   17 |   upper[g] = 'A';
      |   ^     ~
../test.cpp:1:1: note: array 'upper' declared here
    1 | char upper[300];
      | ^
6 warnings generated.
```

This pull request is my first change and I would appreciate any comments or suggestions.

### CODE REVIEW

**Comment by wheatman** [comment]
I found the test I would want to modify here (llvm-project/clang/test/Sema/warn-char-subscripts.c)
I will work on updating it

File: clang/test/Sema/warn-char-subscripts.cpp
Comment (shafik) [inline clang/test/Sema/warn-char-subscripts.cpp:36]: Why do we exclude the `signed char` case?

**Review by shafik — COMMENTED** [review]
You need to make sure you tests have newlines at the end

File: clang/test/Sema/warn-char-subscripts.cpp
Comment (wheatman) [reply clang/test/Sema/warn-char-subscripts.cpp:36]: That is an interesting questions, the behavior for that was not changed.  
The current behavior for explicitly signed or unsigned chars is to have no warning, only unspecified chars have warnings.

https://godbolt.org/z/7oc3ET4h4

File: clang/lib/Sema/SemaExpr.cpp
Comment (AaronBallman) [inline clang/lib/Sema/SemaExpr.cpp:6028]:
```suggestion
    std::optional<llvm::APSInt> IntegerContantExpr =
        IndexExpr->getIntegerConstantExpr(getASTContext());
    if (!IntegerContantExpr.has_value() ||
        IntegerContantExpr.value().isNegative())
      Diag(LLoc, diag::warn_subscript_is_char) << IndexExpr->getSourceRange();
```
This performs extra work for each array subscript in the TU, but because it's limited to only array subscripts with a character type, I don't think the compile-time performance hit should be too bad, so this seems reasonable to me.

File: clang/test/Sema/warn-char-subscripts.c
Comment (AaronBallman) [inline clang/test/Sema/warn-char-subscripts.c:89]: These two both surprise me in a C file; `b` is not an integer constant expression in either case, so I would have expected no diagnostic here. The diagnostic is defensible, just a surprise.

File: clang/test/Sema/warn-char-subscripts.cpp
Comment (AaronBallman) [reply clang/test/Sema/warn-char-subscripts.cpp:36]: We rule out explicitly signed or explicitly unsigned `char` type because of `int8_t` and `uint8_t` because those are pretty reasonable to use as array indexes.

**Review by AaronBallman — COMMENTED** [review]
Thank you for this! The changes should come with a release note in `clang/docs/ReleaseNotes.rst` so users know about the change in behavior.

File: clang/test/Sema/warn-char-subscripts.c
Comment (wheatman) [reply clang/test/Sema/warn-char-subscripts.c:89]: This behavior was not changed https://godbolt.org/z/4drrx6h9M since in `c`  mode `b` is not considered a constant expression.  These tests were added to since they were added in the `c++` file and I wanted to show the expected behavior in both languages.

**Comment by wheatman** [comment]
Thanks you for the comments. I made the updates and added the note in `clang/docs/ReleaseNotes.rst`

**Review by AaronBallman — APPROVED** [review]
LGTM, thank you for the fix!

**Comment by wheatman** [comment]
Thank you for the review.
I don't have write access, so unless @shafik has more comments, could one of you merge it in

**Comment by cor3ntin** [comment]
@wheatman can you rebase again? I want to make sure CI passes before merging (and the bots were broken earlier today). Thanks.
LGTM otherwise

**Comment by cor3ntin** [comment]
Thanks for your first contribution. Maybe the first of many :D

**Pull request merged**
